server/src/main/java/org/apache/cloudstack/storage/fileshare/FileShareServiceImpl.java

```diff
@@ -0,0 +1,687 @@
+
+public class FileShareServiceImpl extends ManagerBase implements FileShareService, Configurable {
+
+    @Inject
+    private AccountManager accountMgr;
+
+    @Inject
+    private EntityManager entityMgr;
+
+    @Inject
+    private ConfigurationManager configMgr;
+
+    @Inject
+    private VolumeApiService volumeApiService;
+
+    @Inject
+    private FileShareDao fileShareDao;
+
+    @Inject
+    private FileShareJoinDao fileShareJoinDao;
+
+    @Inject
+    private DiskOfferingDao diskOfferingDao;
+
+    @Inject
+    ConfigurationDao configDao;
+
+    @Inject
+    VolumeDao volumeDao;
+
+    protected List<FileShareProvider> fileShareProviders;
+
+    private Map<String, FileShareProvider> fileShareProviderMap = new HashMap<>();
+
+    private final StateMachine2<State, Event, FileShare> fileShareStateMachine;
+
+    ScheduledExecutorService _executor = null;
+
+    public FileShareServiceImpl() {
+        this.fileShareStateMachine = State.getStateMachine();
+    }
+
+    @Override
+    public boolean start() {
+        fileShareProviderMap.clear();
+        for (final FileShareProvider provider : fileShareProviders) {
+            fileShareProviderMap.put(provider.getName(), provider);
+            provider.configure();
+        }
+        _executor.scheduleWithFixedDelay(new FileShareGarbageCollector(), FileShareCleanupInterval.value(), FileShareCleanupInterval.value(), TimeUnit.SECONDS);
+        return true;
+    }
+
+    public boolean stop() {
+        _executor.shutdown();
+        return true;
+    }
+
+    @Override
+    public List<FileShareProvider> getFileShareProviders() {
+        return fileShareProviders;
+    }
+
+    @Override
+    public boolean stateTransitTo(FileShare fileShare, Event event) {
+        try {
+            return fileShareStateMachine.transitTo(fileShare, event, null, fileShareDao);
+        } catch (NoTransitionException e) {
+            logger.debug(String.format("Failed during event % for File Share %s [%s] due to exception %",
+                    event.toString(), fileShare.getName(), fileShare.getId(), e));
+            return false;
+        }
+    }
+
+    @Override
+    public void setFileShareProviders(List<FileShareProvider> fileShareProviders) {
+        this.fileShareProviders = fileShareProviders;
+    }
+
+    @Override
+    public FileShareProvider getFileShareProvider(String fileShareProviderName) {
+        if (fileShareProviderMap.containsKey(fileShareProviderName)) {
+            return fileShareProviderMap.get(fileShareProviderName);
+        }
+        throw new CloudRuntimeException("Invalid file share provider name!");
+    }
+
+    @Override
+    public boolean configure(final String name, final Map<String, Object> params) throws ConfigurationException {
+        Map<String, String> configs = configDao.getConfiguration("management-server", params);
+        String workers = configs.get("expunge.workers");
+        int wrks = NumbersUtil.parseInt(workers, 10);
```

Comment:
should this be configurable as well?

Reply:
yes, I can add a new fileshare specific fileshare.expunge.workers setting